perf/src/recycler.rs

```diff
@@ -11,36 +18,128 @@ struct RecyclerStats {
     max_gc: AtomicUsize,
 }
 
-#[derive(Clone, Default)]
-pub struct Recycler<T> {
+#[derive(Debug, Default)]
+struct RecyclerShrinkStats {
+    resulting_size: u32,
+    target_size: u32,
+    ideal_num_to_remove: u32,
+    shrink_elapsed: u64,
+    drop_elapsed: u64,
+}
+
+impl RecyclerShrinkStats {
+    fn report(&self, shrink_metric_name: &'static str) {
+        datapoint_info!(
+            shrink_metric_name,
+            ("target_size", self.target_size as i64, i64),
+            ("resulting_size", self.resulting_size as i64, i64),
+            ("ideal_num_to_remove", self.ideal_num_to_remove as i64, i64),
+            ("recycler_shrink_elapsed", self.shrink_elapsed as i64, i64),
+            ("drop_elapsed", self.drop_elapsed as i64, i64)
+        );
+    }
+}
+
+#[derive(Clone)]
+pub struct Recycler<T: Reset> {
     recycler: Arc<RecyclerX<T>>,
+    shrink_metric_name: &'static str,
+}
+
+impl<T: Default + Reset> Recycler<T> {
+    pub fn new(shrink_metric_name: &'static str) -> Self {
```

Comment:
nits: `new_without_limit`? This looks like it's pretty dangerous to instantiate this if confronted to untrusted data flow. So indicate the name as so. :)